### PR TITLE
Prometheus: Fix Azure authentication support

### PR DESCRIPTION
**Pull request by kostrse**
**What this PR does / why we need it**:

Re-adding back Azure authentication support to Prometheus datasource after the datasource query logic was rewritten from `plugin.json` routes to Go backend.

**Which issue(s) this PR fixes**:

Related #35857 

**Special notes for your reviewer**:

Azure credentials being set to `opts.CustomOptions` when it configured by a user.
Azure credentials currently stored in `opts.CustomOptions` because `httpclient.Options` doesn't have yet strongly-typed fields for Azure configuration but `httpclient.Provider` already has Azure middleware and understands credentials from the `CustomOptions` map (under feature flag `httpclientprovider_azure_auth`).

I need `JsonData` as a dynamic map of values rather than a strongly-typed struct, so it could be passed to `azcredentials.FromDatasourceData` without Prometheus datasource knowing about schema of Azure credentials.

I declared it as `map[string]insterface{}` and added helper methods for access to values stored in the untyped map (`maputil` package). I could use `simplejson` instead of it. Feedback appreciated.

It would be useful to have `JsonData` in Grafana SDK `backend.DataSourceInstanceSettings` as an already deserialized `map[string]insterface{}`, similar to `DecryptedSecureJSONData`, not just only as a raw byte array.

### CODE REVIEW

**Review by marefr — CHANGES_REQUESTED** [review]
This works, but yeah it feels a bit messy moving from types to generic map. Now I think it's time to add the azure credentials configuration to grafana-plugin-sdk-go according to my earlier https://github.com/grafana/grafana/pull/39209#discussion_r783863830 to clean this up. Let us know if you need help with that.

File: pkg/tsdb/prometheus/promclient/provider.go [hunk 1]
```diff
@@ -77,13 +83,31 @@ func (p *Provider) middlewares() []sdkhttpclient.Middleware {
...
 
 	return middlewares
 }
 
+func (p *Provider) configureAzureAuthentication(opts sdkhttpclient.Options) error {
+	credentials, err := azcredentials.FromDatasourceData(p.jsonData, p.settings.DecryptedSecureJSONData)
+	if err != nil {
+		err = fmt.Errorf("invalid Azure credentials: %s", err)
+		return err
+	} else if credentials != nil {
```
[the remark this answers is not in the document]
Comment (kostrse) [reply pkg/tsdb/prometheus/promclient/provider.go]: Fixed this

**Comment by kostrse** [comment]
> This works, but yeah it feels a bit messy moving from types to generic map.

The generic map itself isn't that messy in my opinion, it allows different component to extract necessary information without having to know about each other (each of them may have own strongly-typed model but schema is too simple in this particular case to do that). Another question is how to access the values in a not messy way.

Is it OK to keep the helper methods which I introduced, or I can switch to `simplejson.Json` instead of the `map[string]interface{}`

> Now I think it's time to add the azure credentials configuration to grafana-plugin-sdk-go according to my earlier [#39209 (comment)](https://github.com/grafana/grafana/pull/39209#discussion_r783863830) to clean this up. Let us know if you need help with that.

I will send one more cleanup PR before adding strongly-typed configuration to `grafana-plugin-sdk-go`

**Comment by kostrse** [comment]
@marefr @daniellee @jongyllen this PR fixes Azure authentication in Prometheus datasource. Would very want to have it merged and be included into Grafana 8.4 release. Otherwise we won't be able to upgrade from Grafana 8.2.

**Comment by marefr** [comment]
@kostrse yes understood and we're aware of this. I'll have a new look on this during the day.

**Review by marefr — APPROVED** [review]
LGTM. Just need to see if the build passes, I've restarted it

> > This works, but yeah it feels a bit messy moving from types to generic map.
> 
> The generic map itself isn't that messy in my opinion, it allows different component to extract necessary information without having to know about each other (each of them may have own strongly-typed model but schema is too simple in this particular case to do that). Another question is how to access the values in a not messy way.

Since this is temporary and will change with moving things to the SDK it's fine

> Is it OK to keep the helper methods which I introduced, or I can switch to `simplejson.Json` instead of the `map[string]interface{}`

Since this is temporary and will change with moving things to the SDK, either works.

> > Now I think it's time to add the azure credentials configuration to grafana-plugin-sdk-go according to my earlier [#39209 (comment)](https://github.com/grafana/grafana/pull/39209#discussion_r783863830) to clean this up. Let us know if you need help with that.
> 
> I will send one more cleanup PR before adding strongly-typed configuration to `grafana-plugin-sdk-go`

Sounds good :+1:

**Comment by marefr** [comment]
Thanks @kostrse

**Pull request merged**